mobile_verifier/src/promotion_reward.rs

```diff
@@ -0,0 +1,644 @@
+        settings: &Settings,
+        file_upload: FileUpload,
+        file_store: file_store::FileStore,
+        gateway_info_resolver: GatewayClient,
+        authorization_verifier: AuthorizationClient,
+        entity_verifier: EntityClient,
+    ) -> anyhow::Result<impl ManagedTask> {
+        let (promotion_rewards_sink, valid_promotion_rewards_server) =
+            VerifiedPromotionRewardV1::file_sink(
+                settings.store_base_path(),
+                file_upload.clone(),
+                Some(Duration::from_secs(15 * 60)),
+                env!("CARGO_PKG_NAME"),
+            )
+            .await?;
+
+        let (promotion_rewards, promotion_rewards_server) =
+            file_source::continuous_source::<PromotionReward, _>()
+                .state(pool.clone())
+                .store(file_store)
+                .lookback(LookbackBehavior::StartAfter(settings.start_after))
+                .prefix(FileType::PromotionRewardIngestReport.to_string())
+                .create()
+                .await?;
+
+        let promotion_reward_daemon = Self {
+            pool,
+            gateway_info_resolver,
+            authorization_verifier,
+            entity_verifier,
+            promotion_rewards,
+            promotion_rewards_sink,
+        };
+
+        Ok(TaskManager::builder()
+            .add_task(valid_promotion_rewards_server)
+            .add_task(promotion_rewards_server)
+            .add_task(promotion_reward_daemon)
+            .build())
+    }
+
+    async fn run(mut self, shutdown: triggered::Listener) -> anyhow::Result<()> {
+        loop {
+            tokio::select! {
+                _ = shutdown.clone() => {
+                    tracing::info!("PromotionRewardDaemon shutting down");
+                    break;
+                }
+                Some(file) = self.promotion_rewards.recv() => {
+                    let start = Instant::now();
+                    self.process_file(file).await?;
+                    metrics::histogram!("promotion_reward_processing_time")
+                        .record(start.elapsed());
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn process_file(&self, file: FileInfoStream<PromotionReward>) -> anyhow::Result<()> {
+        tracing::info!("Processing promotion reward file {}", file.file_info.key);
+
+        let mut transaction = self.pool.begin().await?;
+        let reports = file.into_stream(&mut transaction).await?;
+
+        let mut verified_promotion_rewards =
+            pin!(ValidatedPromotionReward::validate_promotion_rewards(
+                reports,
+                &self.authorization_verifier,
+                &self.gateway_info_resolver,
+                &self.entity_verifier
+            ));
+
+        while let Some(promotion_reward) = verified_promotion_rewards.try_next().await? {
+            promotion_reward.write(&self.promotion_rewards_sink).await?;
+            if promotion_reward.is_valid() {
```

Comment:
let see if we can validate here instead of line 122 so it will make code simpler to read